presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/RemoveUnsupportedDynamicFilters.java

```diff
@@ -0,0 +1,126 @@
+import io.prestosql.sql.planner.plan.SimplePlanRewriter;
+import io.prestosql.sql.planner.plan.TableScanNode;
+import io.prestosql.sql.tree.Expression;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.google.common.collect.Maps.filterKeys;
+import static io.prestosql.sql.DynamicFilters.extractDynamicFilters;
+import static io.prestosql.sql.ExpressionUtils.combineConjuncts;
+import static io.prestosql.sql.tree.BooleanLiteral.TRUE_LITERAL;
+
+/**
+ * Dynamic filters are supported only right after TableScan
+ */
+public class RemoveUnsupportedDynamicFilters
+        implements PlanOptimizer
+{
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, TypeProvider types, SymbolAllocator symbolAllocator, PlanNodeIdAllocator idAllocator, WarningCollector warningCollector)
+    {
+        return SimplePlanRewriter.rewriteWith(new RemoveUnsupportedDynamicFilters.Rewriter(), plan, new HashSet<>());
+    }
+
+    private class Rewriter
+            extends SimplePlanRewriter<Set<String>>
+    {
+        @Override
+        public PlanNode visitJoin(JoinNode node, RewriteContext<Set<String>> context)
+        {
+            JoinNode join = (JoinNode) context.defaultRewrite(node, context.get());
+
+            Optional<Expression> filterOptional = join.getFilter().map(filter -> {
+                Expression modified = removeDynamicFilters(filter, context.get());
+                if (TRUE_LITERAL.equals(modified)) {
+                    return null;
+                }
+                else if (!filter.equals(modified)) {
+                    return modified;
+                }
+                return filter;
+            });
+
+            Map<String, Symbol> dynamicFilters = ImmutableMap.copyOf(filterKeys(node.getDynamicFilters(), key -> !context.get().contains(key)));
+
+            if (!filterOptional.equals(join.getFilter()) || !dynamicFilters.equals(join.getDynamicFilters())) {
+                return new JoinNode(
+                        join.getId(),
+                        join.getType(),
+                        join.getLeft(),
+                        join.getRight(),
+                        join.getCriteria(),
+                        join.getOutputSymbols(),
+                        filterOptional,
+                        join.getLeftHashSymbol(),
+                        join.getRightHashSymbol(),
+                        join.getDistributionType(),
+                        join.isSpillable(),
+                        dynamicFilters);
+            }
+            return join;
+        }
+
+        @Override
+        public PlanNode visitFilter(FilterNode node, RewriteContext<Set<String>> context)
+        {
+            FilterNode filter = (FilterNode) context.defaultRewrite(node, context.get());
+            PlanNode source = filter.getSource();
+            if (source instanceof TableScanNode) {
+                return filter;
```

Comment:
this still doesn't remove node filters from build side.

Reply:
I've moved out `RemoveUnsupportedDynamicFilters` and `DynamicFiltersChecker` classes into the 2nd commit to avoid confusion.